service/labelstore/service.go

```diff
@@ -196,26 +198,23 @@ func (s *service) GetLocalRefID(componentID string, globalRefID uint64) uint64 {
 	return local
 }
 
-// AddStaleMarker adds a stale marker
-func (s *service) AddStaleMarker(globalRefID uint64, l labels.Labels) {
+func (s *service) TrackStaleness(ids []StalenessTracker) {
 	s.mut.Lock()
 	defer s.mut.Unlock()
 
-	s.staleGlobals[globalRefID] = &staleMarker{
-		lastMarkedStale: time.Now(),
-		labelHash:       l.Hash(),
-		globalID:        globalRefID,
+	for _, id := range ids {
+		if value.IsStaleNaN(id.Value) {
+			s.staleGlobals[id.GlobalRefID] = &staleMarker{
```

Comment:
Here we have a map of pointers, but in the fanout we use a slice of structs `stalenessTrackers []labelstore.StalenessTracker`. We also use slice of structs for labels. So it's a bit inconsistent and I'm not sure what's the thinking behind it. Do we have benchmarks for what performs better? Did this come up in allocation profiles somewhere?

Reply:
You dont have the traverse the map to check if you need to add or update the value like you would an array. Now its unlikely you would need to add staleness markers mutliple times, but I would rather not assume that.